### PR TITLE
Implementation of Calcjobs, Parsers, and entry points for the XSpectra post-processing code

### PR DESCRIPTION
**Pull request by PNOGillespie**
## Request Overview

Adds CalcJob and Parser classes, as well as associated entry points, to run the [XSpectra](https://www.quantum-espresso.org/Doc/INPUT_XSpectra.txt) post-processing code provided with distributions of Quantum ESPRESSO. The proposed additions would add the following:

* CalcJob classes for xspectra.x and its associated upf2plotcore.sh utility script
* Parser classes for xspectra.x and upf2plotcore.sh
* Entry points for both CalcJobs and Parsers in the head pyproject.toml file

The XSpectra parts function the same as many other post-processing codes, with some differences (as mentioned later on), while the UPF2Plotcore parts run a utility script provided with each distribution of XSpectra. This script collects the radial part of the core wavefunction from a provided pseudopotential file that contains GIPAW information and returns the plaintext data in a SingleFile node which is needed for the XSpectra calculation (see later for more information).

The additions have been extensively tested in private by our group at CNR-NANO prior to upload and - since XSpectra is a spectrum-generating post-processing code for QE at its core - makes use of the python code for the dos.x plugin as a basis before being modified to work with XSpectra. As this is my first contribution to a public GitHub repository, I would like to apologise in advance if I have missed any best practice steps along the way.

## XSpectra Overview

In order to add some context to the submission, I provide here a brief overview of the XSpectra code itself.

The XSpectra code computes X-ray absorption spectra (XAS), based on the theory developed by [Mauri *et al* \(2002\)](https://journals.aps.org/prb/abstract/10.1103/PhysRevB.66.195107), from a converged SCF wavefunction in a similar manner to other post-processing (PP) codes. Where it differs from other PP codes is that the calculation of the XAS cross-section requires wavefunctions for both the initial and final-states of the excitation.

The final-state wavefunction is taken from a converged SCF calculation which, In order to improve the description of the final-state, often involves performing the SCF calculation where the atom chosen as the absorbing atom is assigned a pseudopotential with a hole in a given core level.

The initial state wavefunction is taken from the upf2plotcore.sh script, which takes a upf file for a pseudopotential as an input, re-constructs the core wavefunction, and outputs a plaintext file containing the radial part of the core wavefunction. If the pseudopotential does not contain GIPAW information, the code simply prints a plaintext file with no wavefunction data.

The calculation steps are therefore:
* Run an SCF calculation using pw.x for the system of interest, typically with a core-hole pseudopotential file for the chosen absorbing atom.
* Collect the initial-state wavefunction data using upf2plotcore.sh and a pseudopotential for the absorbing element type which contains GIPAW information.
* Run xspectra.x using the initial and final states as inputs and collect the spectrum data file as an output.

### CODE REVIEW

**Comment by sphuber** [comment]
Thanks @PNOGillespie , it would be great to have support for `xspectra.x`. Before going into a more detailed review though, I wanted to first discuss the approach. The `xspectra.x` is a real code and so using a `CalcJob` plugin as `XSpectraCalculation` is great. However, I am wondering if it is not overkill for the `ufp2plotcore.sh` plot script. Currently it is indeed required for AiiDA to write a dedicated plugin and parser, but I recently wrote the [`aiida-shell` plugin](https://github.com/sphuber/aiida-shell/). This makes it dead easy to run simple shell scripts/commands like this without needing to write a plugin, but with all the advantages of automatic provenance.

It would look something like
```python
from aiida_shell import launch_shell_job
upf = load_node(pk)  # This would also used in the input of the `PwCalculation`.
results, node = launch_shell_job(
    'upf2plotcore.sh',
    arguments=['{upf}'],
    files={'upf': upf},
    metadata={'options': {'computer': load_computer('some-computer')}}
)
inputs = {
    'code': load_code('xspectra@machine'),
    'core_wfc_data': results['stdout'],
    'parent_folder': load_node(pk).outputs.remote_folder,
    'bz_sampling': KpointsData(),
}
submit(XspectraCalculation, **inputs)
```
It would be great if you could give that a shot sometime

**Comment by PNOGillespie** [comment]
Thank you for the response @sphuber. I originally wrote the `upf2plotcore.sh` CalcJob plugin as a first step to understand how the CalcJob class works in practice before moving to `xspectra.x` (being as it is a simple example of input text -> run CJ -> parse text -> save to node) however I agree that the CJ implementation is a heavy-handed way to run it, as the time taken for AiiDA to process the job is much longer than the runtime of the script itself! For testing with a simple CJ implementation of `upf2plotcore` this wasn't too much of an issue, since the job would only be run once for each GIPAW pseudo and then the SingleFileData output would be saved somewhere for later use, but now that we're looking to develop formal workchains for XSpectra I can see it becoming unwieldy if it needs to be executed each time the workchain runs. 

I didn't know about `aiida-shell` before, but I agree that it looks like a simpler way to get the same data. I will take a look at it this week and see how it goes.

**Comment by sphuber** [comment]
Great @PNOGillespie ! Note that the runtime of the `upf2plotcore` job may have been "artificially" long. Where you running this on the `localhost` computer? If so, would be good to check in a `verdi` shell what `load_computer(PK).get_minimum_job_poll_interval()`. My bet is that this would show some value between 2 and 10. If that is the case, run `load_computer(PK).get_minimum_job_poll_interval(0.1)`. This should make the jobs run a lot faster as it is not arbitrarily waiting anymore. The interval is not that important for the `localhost` computer.

**Comment by PNOGillespie** [comment]
Hi @sphuber, apologies for the slow response. I had a go with `aiida-shell` and I think it would work well as a replacement for a calcjob implementation for upf2plotcore - as you say, `aiida-shell` is much easier to work with and does everything that the calcjob class would have done anyway.

I will remove the UPF2PlotcoreCalculation parts from the request and wait for a full review of the rest of the code.

File: src/aiida_quantumespresso/parsers/xspectra.py
Comment (sphuber) [inline src/aiida_quantumespresso/parsers/xspectra.py:86]: You are not catching these exceptions right? Would probably be best to catch these in the `parse` method and convert to an exit code instead of letting the parser except.

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @PNOGillespie I have given it a first pass. There are few more high-level comments:

* It seems you haven't run the pre-commit which automatically formats and lints the code. Please install the repository with `pip install -e .[pre-commit]` and then run `pre-commit install`. After that, each time you commit, the linters will automatically run. You can also run it manually with `pre-commit run --all`.
* There are no tests and it would be good to have at least a test for a simple case for both the `CalcJob` and `Parser` implementations. You can look at the other tests how that is done but I can also help if you have difficulty with it. Essentially we would need the output file of a very simple xspectra.x calculation.
* You parse the output from the stdout file, which is quite fragile. Does xspectra by any chance also write to an XML file like many other QE codes?

File: src/aiida_quantumespresso/calculations/xspectra.py
Comment (PNOGillespie) [reply src/aiida_quantumespresso/calculations/xspectra.py:41]: This was a deliberate choice from the beginning in order to clearly differentiate the k-point mesh used by PwSCF and XSpectra when running python scripts. This was done since we believed early on in the process that the convergence of the XAS spectrum w/r/t kpoints would likely be different than the total energy - and that likely we would need to establish a protocol to converge this parameter.

The original idea I had was to use a different name to avoid confusion in the inputs when this is implemented in a full workchain. However if it would be more consistent, and isn't too difficult to work with in a full workchain, then I will go with your suggestion.

**Comment by PNOGillespie** [comment]
Thank you for your review @sphuber and your helpful comments - apologies again for the various mistakes, though I suppose that I'm learning a lot during this first submission process!

>    It seems you haven't run the pre-commit which automatically formats and lints the code. Please install the repository with pip install -e .[pre-commit] and then run pre-commit install. After that, each time you commit, the linters will automatically run. You can also run it manually with pre-commit run --all.

I will download and install this sometime soon, though I have some calculation work running at the moment which I will have to wait for first before proceeding.

>    There are no tests and it would be good to have at least a test for a simple case for both the CalcJob and Parser implementations. You can look at the other tests how that is done but I can also help if you have difficulty with it. Essentially we would need the output file of a very simple xspectra.x calculation.

I can think of a few examples to use from my own testing, so I will add these when I get the chance this week.

>    You parse the output from the stdout file, which is quite fragile. Does xspectra by any chance also write to an XML file like many other QE codes?

Unfortunately, the code does not produce an XML file. The files it produces are the stdout, spectrum file, and a .save file used for re-starts which I intend to write a procedure for in the near future.

**Comment by PNOGillespie** [comment]
Hi @sphuber I have added two commits to address the various suggested changes to the `CalcJob` and `Parser` classes:

[XspectraCalculation](https://github.com/aiidateam/aiida-quantumespresso/pull/815/commits/a7390e34f1fac43ff2b7647999a9ad9a1de57d96):
* The kpoint mesh input has been modified as you suggested. This method is essentially the same as the one which the `BasePwCpInputGenerator` class uses, including the few lines which it uses to ensure that the offset is ok.
* The `prepare_for_submission()` method has been simplified significantly: since all that needs to be done here is to add the core wavefunction data file to the `local_copy_list`, the method now simply overrides the inherited method and appends the requisite node. Originally, this method was just a copy of an older version of the NamelistsCalculation class (hence why the code seemed out of date) but now I hope it should follow the changes made to the parent class going forward.

[XspectraParser](https://github.com/aiidateam/aiida-quantumespresso/pull/815/commits/3c83c4d9bc62d3bc7f752d9add509fe001e0a8be)
* The redundant loop that you spotted has been removed - this same check was already performed in the main `parse` method, so I simply deleted the duplicate
* The check for `np.genfromtxt()` has been moved to the main `parse` method as well, and I have added an error code for it as you suggested.
* I realised that I had neglected to include lines in the parser for handling spin-polarised calculations, so I have extended the Parser to handle this as well.

Regarding the tests: from what I understand of the tests directory, this would just involve providing Parser tests - for which I need to:
* Add a folder under `tests/parsers/fixtures/xspectra` containing the pre-generated output files for each test
* Add a testing script to `tests/parsers`
* Add a set of already-parsed outputs to `tests/parsers/test_xspectra` to compare against

Please let me know if I have missed anything out in that. I think the tests provided for `dos.x` should provide useful inspiration for how to do this properly. I will work on adding the tests and hopefully commit those in the next day or two if all goes well.

Many thanks again for all the help and suggestions.

**Comment by PNOGillespie** [comment]
Ok @sphuber, I think the code is ready for a second review. [As stated before](https://github.com/aiidateam/aiida-quantumespresso/pull/815#issuecomment-1151342523) the `XspectraCalculation` class now simply calls the inherited `prepare_for_submission()` method and then just appends the core wavefunction data file to the `local_copy_list` - thus greatly simplifying things. The `XspectraParser` class now correctly distinguishes between spin-polarised and non-spin-polarised calculations and reports all the relevant information in each case. Finally, I have added tests for successful runs both with and without spin-polarisation, as well as others for an interrupted calculation and the two error codes specific for XSpectra (313 and 314).

Thanks in advance.

**Review by sphuber — APPROVED** [review]
Thanks a lot @PNOGillespie this is a great pull request and great addition for the plugin. Especially want to thank you for your patience in responding to requested changes and adding the tests :+1:  Only thing missing is a bit of documentation, especially since the workflow would require using the `aiida-shell` plugin to run the `upf2core.sh` script, but that can be added later.

**Comment by PNOGillespie** [comment]
Many thanks @sphuber! I should also thank you for all of your assistance along the way. Work is already underway on a WorkChain, which I intend to write some documentation for once I have a basic version completed, so expect more XSpectra-related content in the near-future.

**Pull request merged**
